### PR TITLE
[GEOT-7629] Add validation when deserializing objects

### PR DESCRIPTION
**Pull request by sikeoka**
[GEOT-7629](https://osgeo-org.atlassian.net/browse/GEOT-7629) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)
This PR uses ValidatingObjectInputStream from the Commons IO library to to restrict the classes that are allowed when deserializing imagemosaic sample image and histogram files and graph files. There are multiple existing unit tests that test deserializing valid graph files. The sample image files are the primary concern for potential backwards-compatibility issues.

SimpleFeatureIO was restricted to only allow deserialization when the constructor is called with an empty file with a system property to completely enable or disable deserialization. This allows the existing use cases where a new temp file is created and a single SimpleFeatureIO instance will serialize data to the file and then deserialize the data back from the file.

An unused method in the Base64 class was deprecated with a warning to discourage people from using it.

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/library/main/src/main/java/org/geotools/data/sort/SimpleFeatureIO.java [hunk 1]
```diff
@@ -269,6 +276,12 @@ private byte[] getByteBuffer(int length) {
 
     @SuppressWarnings("BanSerializableRead")
     private Object readObject() throws IOException {
+        if ("false".equalsIgnoreCase(this.enableDeserialization)) {
+            throw new IllegalStateException("Object deserialization is not allowed");
+        } else if (!this.initialFileEmpty && !"true".equalsIgnoreCase(this.enableDeserialization)) {
+            throw new IllegalStateException(
+                    "Object deserialization is only allowed when created with an empty file");
+        }
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/data/sort/SimpleFeatureIO.java:284]: SimpleFeatureIO is used to perform sort/merge over secondary memory (when past a certain limit of features, below it works in memory). This is the default behavior of ContentDataStore, so all stores inherit it unless they can provide native sorting. 

Now, most attributes will be simple types and won't be affected, but it's not uncommon for stores nowadays to add Map objects as attributes to support more complex types. For example, PostgreSQL support for hstore and JSON uses maps (likely saved by native order by encoding), and I'm guessing ElasticSearch and Mongo too (but they also support native sorting).

Long story short, while I cannot find an example right away, the chance of breaking some existing implementation out there is not insignificant. This should probably be marked in the upgrade notes for both GeoTools and GeoServer.

File: modules/library/metadata/src/main/java/org/geotools/util/Base64.java [hunk 2]
```diff
@@ -898,7 +898,9 @@ public static byte[] decode(String s) {
      * @param encodedObject The Base64 data to decode
      * @return The decoded and deserialized object
      * @since 1.5
+     * @deprecated This method is not secure and will be removed in a future release.
```
Comment (aaime) [inline modules/library/metadata/src/main/java/org/geotools/util/Base64.java:901]: Doesn't seem to be used anyways, so looks like a safe removal

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/Utils.java [hunk 3]
```diff
@@ -236,6 +235,25 @@ public class Utils {
...
+        float[].class,
+        int[].class,
+        short[].class,
+        java.math.BigInteger.class,
+        java.net.InetAddress.class,
+        java.util.Hashtable.class
+    };
+
+    /** Patterns for qualified class names to allow when deserializing SampleImage objects */
+    private static final String[] SAMPLE_IMAGE_PATTERNS = {
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/Utils.java:253]: Yep, you stumbled into the reason why serialization was used in the first place. There are so many possible combinations of sample and color models that it's not practical to try and enumerate them all.

This has some potential to break existing implementations that use custom-written image readers. The patterns here should be made configurable with a system variable.

**Review by aaime — CHANGES_REQUESTED** [review]
Looks good, seem comment below

File: modules/library/main/src/main/java/org/geotools/data/sort/SimpleFeatureIO.java (hunk 1 above)
Comment (sikeoka) [reply modules/library/main/src/main/java/org/geotools/data/sort/SimpleFeatureIO.java:284]: There are two existing cases where SimpleFeatureIO is used and both create a single SimpleFeatureIO object on an empty temporary file and then write to and read from the file from the same object. The default behavior was specifically chosen to allow this behavior.
https://github.com/geotools/geotools/blob/main/modules/library/main/src/main/java/org/geotools/data/sort/MergeSortDumper.java#L137
https://github.com/geotools/geotools/blob/main/modules/library/render/src/main/java/org/geotools/renderer/lite/MarkFeatureIterator.java#L161

Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/data/sort/SimpleFeatureIO.java:284]: Isn't "enableDeserialization" is false by default? Then the second if won't be reached at all... Oh... by default it's the empty string... that's why I found it confusing, normally for boolean system variables we use Boolean.getBoolean instead.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/Utils.java (hunk 3 above)
Comment (sikeoka) [reply modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/Utils.java:254]: I added a property to allow administrators to allow additional classes for sample image files.

Comment (aaime) [reply modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/Utils.java:254]: Thanks, that helps. It should be also noted in the upgrade notes (both the stricter checks and the system variable to allow other classes be deserialized).

**Comment by sikeoka** [comment]
I created https://github.com/geoserver/geoserver/pull/7855 to add the new property to the GeoServer upgrade notes.

**Comment by jodygarnett** [comment]
Are we comfortable back porting this one? I guess we are ..
